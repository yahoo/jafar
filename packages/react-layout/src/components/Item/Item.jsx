/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import ItemView from '../ItemView';
import MobileItemView from '../MobileItemView';

export const layoutTypes = {
  TABS: 'tabs',
  SCROLL: 'scroll',
  MOBILE: 'mobile',
};
export default class Item extends React.Component {
  static propTypes = {
    title: PropTypes.string,
    layout: PropTypes.oneOf(Object.values(layoutTypes)),
    sections: PropTypes.array.isRequired,
    selected: PropTypes.object,
    mainActions: PropTypes.array,
    optionsActions: PropTypes.array,
    size: PropTypes.number,
  };

  constructor(props) {
    super(props);

    this.state = {
      selected: props.selected || {},
    };
  }

  render() {
    // filter section by exclude condition
    const filteredSections = filterSectionsByExcludeCondition(this.props.sections);

    if (filteredSections.length === 0) {
      return null;
    }

    const mainActions = this.props.mainActions ? prepareActions(this.props.mainActions) : undefined;
    const optionsActions = this.props.optionsActions ? prepareActions(this.props.optionsActions) : undefined;

    switch (this.props.layout) {
    case layoutTypes.MOBILE: {
      return (<MobileItemView title={this.props.title} sections={filteredSections} mainActions={mainActions}
        optionsActions={optionsActions} size={this.props.size} />);
    }
    default: {
      const selectedSectionId = this.state.selected.sectionId || filteredSections[0].id;

      // create the tabs menu items from the sections structure
      const tabs = this.props.layout ? {
        items: filteredSections.map(section => ({ value: section.id, label: section.title })),
        value: selectedSectionId,
        onChange: this.onChangeTab,
      } : undefined;

      this.mainActionsRefs = this.mainActionsRefs || {};
      (mainActions || []).forEach((action) => {
        if (action.popover) {
          if (!this.mainActionsRefs[action.label]) {
            this.mainActionsRefs[action.label] = React.createRef();
          }
          action.elementRef = this.mainActionsRefs[action.label];
          action.popover.targetRef = this.mainActionsRefs[action.label];
        }
      });

      this.sectionsRef = React.createRef();
      this.sectionsRefs = {};
      filteredSections.forEach((section) => {
        this.sectionsRefs[section.id] = React.createRef();
      });

      // prepare the current displayed sections and add ref to sections
      let sections = filteredSections.map(section => ({ ...section, ref: this.sectionsRefs[section.id] }));

      // if tabs layout - show only selected tab, without title for first level section
      if (this.props.layout === layoutTypes.TABS) {
        sections = [{ ...sections.find(section => section.id === selectedSectionId), title: undefined }];
      }

      // prepare footer
      const footer = this.props.mainActions ? { actions: mainActions } : undefined;

      // prepare options
      const options = this.props.optionsActions ? { actions: optionsActions } : undefined;

      return (<ItemView title={this.props.title} tabs={tabs} sections={sections} sectionsRef={this.sectionsRef}
        footer={footer} options={options} size={this.props.size} />);
    }
    }
  }

  static getDerivedStateFromProps(props, state) {
    if (props.selected && props.selected !== state.selected && props.selected !== state.prevSelected) {
      return ({
        needScroll: true,
        selected: props.selected,
        prevSelected: props.selected, // react doesn't provide prev props for compare here and in componentDidUpdate
      });
    }
    return null;
  }

  componentDidUpdate() {
    if (this.state.needScroll) {
      this.setState({
        needScroll: false,
      });

      // scroll to element
      if (this.state.selected.elementId) {
        const sections = this.sectionsRef.current;
        sections.scrollTo({
          top: sections.querySelector(`#${this.state.selected.elementId}`).offsetTop - sections.offsetTop - 10,
          left: 0,
          behavior: 'smooth',
        });
      }
    }
  }

  onChangeTab = (event, selectedSectionId) => {
    this.setState({
      selected: { sectionId: selectedSectionId, elementId: undefined },
    });

    // if scroll layout - scroll to section
    if (this.props.layout === layoutTypes.SCROLL) {
      this.sectionsRefs[selectedSectionId].current.scrollIntoView({ behavior: 'smooth' });
    }
  }
}

function filterSectionsByExcludeCondition(sections = []) {
  const newSections = sections.map(s => ({ ...s }));
  return newSections.filter((section) => {
    const renderSections = filterSectionsByExcludeCondition(section.sections);
    if (renderSections.length) {
      section.sections = renderSections;
    }
    return !(section.exclude && section.exclude()) || section.sections;
  });
}

function prepareActions(actions) {
  return actions
    .filter(action => !(action.exclude && action.exclude()))
    .map(action => Object.assign({}, action, {
      disabled: action.disable && action.disable(),
    }));
}
