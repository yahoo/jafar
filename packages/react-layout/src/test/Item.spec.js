/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { shallow } from 'enzyme';
import Item from '../components/Item';
 
describe('Item', () => {
  let props;
 
  beforeEach(() => {
    props = {
      title: 'Employee',
      sections: [{
        id: 'general',
        title: 'General',
      }],
      layout: 'scroll',
      mainActions: [],
      optionsActions: [],
    };
  });
  it('Should render - ok', () => {
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });
 
  it('Should render with layout tabs - ok', () => {
    props.layout = 'tabs';
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });
 
  it('Should render with layout mobile - ok', () => {
    props.layout = 'mobile';
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });
 
  it('Should render without layout - ok', () => {
    props.layout = undefined;
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });
 
  it('Should render without title - ok', () => {
    props.title = undefined;
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });
 
  it('Should render with all filtered sections - ok', () => {
    props.sections[0].exclude = () => true;
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });
 
  it('Should render without mainActions - ok', () => {
    props.mainActions = undefined;
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });
 
  it('Should render with mainActions - popover - ok', () => {
    props.mainActions = [{
      label: 'Save & Close',
      type: 'secondary',
      disable: () => false,
      onClick: jest.fn(),
    }, {
      label: 'Save',
      type: 'primary',
      icon: () => null,
      disable: () => false,
      onClick: jest.fn(),
      popover: {
        title: 'Handle Fields',
        open: () => false,
        component: () => (<div>hi</div>),
        props: { onClickField: jest.fn() },
      },
    }];
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });
 
  it('Should render without optionsActions - ok', () => {
    props.optionsActions = undefined;
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });
 
  it('getDerivedStateFromProps - selected defined', () => {
    props.selected = { sectionId: 'general' };
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });
 
  it('getDerivedStateFromProps - return need scroll', () => {
    props.selected = { sectionId: 'general' };
    const state = {};
    const result = Item.getDerivedStateFromProps(props, state);
    expect(result).toEqual({
      needScroll: true,
      selected: props.selected,
      prevSelected: props.selected,
    });
  });
 
  it('onChangeTab - scroll to element', () => {
    props.layout = 'tabs';
    props.selected = { sectionId: 'general', elementId: 'a' };
    const component = shallow(getComponent(props));
    const element = {
      scrollIntoView: jest.fn(),
      querySelector: jest.fn(x => ({ offsetTop: 200 })),
      offsetTop: 150,
    };
    const instance = component.instance();
    instance.state = {
      needScroll: true,
      selected: props.selected,
      prevSelected: props.selected,
    };
    instance.sectionsRefs.general.current = element;
    instance.onChangeTab({}, 'general');
    expect(element.scrollIntoView).not.toHaveBeenCalled();
  });
 
  function getComponent(props) {
    return (<Item {...props} />);
  }
});
 
