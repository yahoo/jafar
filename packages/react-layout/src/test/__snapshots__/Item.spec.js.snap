// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item Should render - ok 1`] = `
<ItemView
  footer={
    Object {
      "actions": Array [],
    }
  }
  options={
    Object {
      "actions": Array [],
    }
  }
  sections={
    Array [
      Object {
        "id": "general",
        "ref": Object {
          "current": null,
        },
        "title": "General",
      },
    ]
  }
  sectionsRef={
    Object {
      "current": null,
    }
  }
  size={4}
  tabs={
    Object {
      "items": Array [
        Object {
          "label": "General",
          "value": "general",
        },
      ],
      "onChange": [Function],
      "value": "general",
    }
  }
  title="Employee"
/>
`;

exports[`Item Should render with all filtered sections - ok 1`] = `""`;

exports[`Item Should render with layout mobile - ok 1`] = `
<MobileItemView
  mainActions={Array []}
  optionsActions={Array []}
  sections={
    Array [
      Object {
        "id": "general",
        "title": "General",
      },
    ]
  }
  size={1}
  title="Employee"
/>
`;

exports[`Item Should render with layout tabs - ok 1`] = `
<ItemView
  footer={
    Object {
      "actions": Array [],
    }
  }
  options={
    Object {
      "actions": Array [],
    }
  }
  sections={
    Array [
      Object {
        "id": "general",
        "ref": Object {
          "current": null,
        },
        "title": undefined,
      },
    ]
  }
  sectionsRef={
    Object {
      "current": null,
    }
  }
  size={4}
  tabs={
    Object {
      "items": Array [
        Object {
          "label": "General",
          "value": "general",
        },
      ],
      "onChange": [Function],
      "value": "general",
    }
  }
  title="Employee"
/>
`;

exports[`Item Should render with mainActions - popover - ok 1`] = `
<ItemView
  footer={
    Object {
      "actions": Array [
        Object {
          "disable": [Function],
          "disabled": false,
          "label": "Save & Close",
          "onClick": [MockFunction],
          "type": "secondary",
        },
        Object {
          "disable": [Function],
          "disabled": false,
          "elementRef": Object {
            "current": null,
          },
          "icon": [Function],
          "label": "Save",
          "onClick": [MockFunction],
          "popover": Object {
            "component": [Function],
            "open": [Function],
            "props": Object {
              "onClickField": [MockFunction],
            },
            "targetRef": Object {
              "current": null,
            },
            "title": "Handle Fields",
          },
          "type": "primary",
        },
      ],
    }
  }
  options={
    Object {
      "actions": Array [],
    }
  }
  sections={
    Array [
      Object {
        "id": "general",
        "ref": Object {
          "current": null,
        },
        "title": "General",
      },
    ]
  }
  sectionsRef={
    Object {
      "current": null,
    }
  }
  size={4}
  tabs={
    Object {
      "items": Array [
        Object {
          "label": "General",
          "value": "general",
        },
      ],
      "onChange": [Function],
      "value": "general",
    }
  }
  title="Employee"
/>
`;

exports[`Item Should render without layout - ok 1`] = `
<ItemView
  footer={
    Object {
      "actions": Array [],
    }
  }
  options={
    Object {
      "actions": Array [],
    }
  }
  sections={
    Array [
      Object {
        "id": "general",
        "ref": Object {
          "current": null,
        },
        "title": "General",
      },
    ]
  }
  sectionsRef={
    Object {
      "current": null,
    }
  }
  size={4}
  title="Employee"
/>
`;

exports[`Item Should render without mainActions - ok 1`] = `
<ItemView
  options={
    Object {
      "actions": Array [],
    }
  }
  sections={
    Array [
      Object {
        "id": "general",
        "ref": Object {
          "current": null,
        },
        "title": "General",
      },
    ]
  }
  sectionsRef={
    Object {
      "current": null,
    }
  }
  size={4}
  tabs={
    Object {
      "items": Array [
        Object {
          "label": "General",
          "value": "general",
        },
      ],
      "onChange": [Function],
      "value": "general",
    }
  }
  title="Employee"
/>
`;

exports[`Item Should render without optionsActions - ok 1`] = `
<ItemView
  footer={
    Object {
      "actions": Array [],
    }
  }
  sections={
    Array [
      Object {
        "id": "general",
        "ref": Object {
          "current": null,
        },
        "title": "General",
      },
    ]
  }
  sectionsRef={
    Object {
      "current": null,
    }
  }
  size={4}
  tabs={
    Object {
      "items": Array [
        Object {
          "label": "General",
          "value": "general",
        },
      ],
      "onChange": [Function],
      "value": "general",
    }
  }
  title="Employee"
/>
`;

exports[`Item Should render without title - ok 1`] = `
<ItemView
  footer={
    Object {
      "actions": Array [],
    }
  }
  options={
    Object {
      "actions": Array [],
    }
  }
  sections={
    Array [
      Object {
        "id": "general",
        "ref": Object {
          "current": null,
        },
        "title": "General",
      },
    ]
  }
  sectionsRef={
    Object {
      "current": null,
    }
  }
  size={4}
  tabs={
    Object {
      "items": Array [
        Object {
          "label": "General",
          "value": "general",
        },
      ],
      "onChange": [Function],
      "value": "general",
    }
  }
/>
`;

exports[`Item getDerivedStateFromProps - selected defined 1`] = `
<ItemView
  footer={
    Object {
      "actions": Array [],
    }
  }
  options={
    Object {
      "actions": Array [],
    }
  }
  sections={
    Array [
      Object {
        "id": "general",
        "ref": Object {
          "current": null,
        },
        "title": "General",
      },
    ]
  }
  sectionsRef={
    Object {
      "current": null,
    }
  }
  size={4}
  tabs={
    Object {
      "items": Array [
        Object {
          "label": "General",
          "value": "general",
        },
      ],
      "onChange": [Function],
      "value": "general",
    }
  }
  title="Employee"
/>
`;
